Exception from Filter is not caught by HttpErrorHandler.onServerError

### Play Version (2.5.x / etc)

2.6.x

### API (Scala / Java / Neither / Both)

Both

### Expected Behavior

Please describe the expected behavior of the issue, starting from the first action.

1. When a `Filter` throws an exception, it should be caught by `HttpErrorHandler.onServerError`.

### Actual Behavior

Please provide a description of what actually happens, working from the same starting point.

Be descriptive: "it doesn't work" does not describe what the behavior actually is -- instead, say "the page renders a 500 error code with no body content."  Copy and paste logs, and include any URLs.  Turn on internal Play logging with `<logger name="play" value="TRACE"/>` if there is no log output.  

1. For Akka HTTP, browser shows 500 error without content. And it logs in the console,
    `[ERROR] [12/11/2017 12:08:11.909] [play-dev-mode-akka.actor.default-dispatcher-6] [akka.actor.ActorSystemImpl(play-dev-mode)] Internal server error, sending 500 response
java.lang.RuntimeException: WRONG
	at filters.ExampleFilter.apply(ExampleFilter.scala:19)
	at play.api.mvc.Filter$$anon$1.apply(Filters.scala:61)
	at play.api.mvc.Filter$$anon$1.apply(Filters.scala:48)
	at play.core.server.AkkaHttpServer.$anonfun$executeHandler$1(AkkaHttpServer.scala:251)
	at play.api.mvc.EssentialAction$$anon$3.apply(Action.scala:48)
	at play.api.mvc.EssentialAction$$anon$3.apply(Action.scala:47)
	at play.core.server.AkkaHttpServer.executeAction(AkkaHttpServer.scala:285)
	at play.core.server.AkkaHttpServer.executeHandler(AkkaHttpServer.scala:255)
	at play.core.server.AkkaHttpServer.handleRequest(AkkaHttpServer.scala:201)
	at play.core.server.AkkaHttpServer.$anonfun$createServerBinding$1(AkkaHttpServer.scala:107)
	at akka.stream.impl.fusing.MapAsync$$anon$24.onPush(Ops.scala:1191)
	at akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:512)
	at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:475)
	at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:371)
	at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:584)
	at akka.stream.impl.fusing.GraphInterpreterShell$AsyncInput.execute(ActorGraphInterpreter.scala:468)
	at akka.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:559)
	at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:741)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:756)
	at akka.actor.Actor.aroundReceive(Actor.scala:517)
	at akka.actor.Actor.aroundReceive$(Actor.scala:515)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:666)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:527)
	at akka.actor.ActorCell.invoke(ActorCell.scala:496)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)`
2. For Netty server, browser just waits for response. And it logs in the console,
    `[error] p.c.s.n.PlayRequestHandler - Exception caught in channelRead future
java.lang.RuntimeException: WRONG
	at filters.ExampleFilter.apply(ExampleFilter.scala:19)
	at play.api.mvc.Filter$$anon$1.apply(Filters.scala:61)
	at play.api.mvc.Filter$$anon$1.apply(Filters.scala:48)
	at play.core.server.netty.PlayRequestHandler.$anonfun$handle$3(PlayRequestHandler.scala:114)
	at play.api.mvc.EssentialAction$$anon$3.apply(Action.scala:48)
	at play.api.mvc.EssentialAction$$anon$3.apply(Action.scala:47)
	at play.core.server.netty.PlayRequestHandler.$anonfun$handleAction$3(PlayRequestHandler.scala:274)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:655)
	at scala.util.Success.$anonfun$map$1(Try.scala:251)
	at scala.util.Success.map(Try.scala:209)`

### Reproducible Test Case

Please provide a PR with a failing test.  

If the issue is more complex or requires configuration, please provide a link to a project on Github that reproduces the issue.

https://github.com/FranklinYinanDing/play-scala-starter-example/tree/2.6.x

--- a/framework/src/play-akka-http-server/src/main/scala/play/core/server/AkkaHttpServer.scala
+++ b/framework/src/play-akka-http-server/src/main/scala/play/core/server/AkkaHttpServer.scala
@@ -31,7 +31,7 @@ import play.core.ApplicationProvider
 import play.server.SSLEngineProvider
 
 import scala.concurrent.duration._
-import scala.concurrent.{ Await, Future }
+import scala.concurrent.{ Await, ExecutionContext, Future }
 import scala.util.control.NonFatal
 import scala.util.{ Failure, Success, Try }
 
@@ -243,23 +243,20 @@ class AkkaHttpServer(
       case Failure(_) => DefaultHttpErrorHandler
     }
 
+    // default execution context used for executing the action
+    implicit val defaultExecutionContext: ExecutionContext = tryApp match {
+      case Success(app) => app.actorSystem.dispatcher
+      case Failure(_) => actorSystem.dispatcher
+    }
+
     (handler, upgradeToWebSocket) match {
       //execute normal action
       case (action: EssentialAction, _) =>
-        val actionWithErrorHandling = EssentialAction { rh =>
-          import play.core.Execution.Implicits.trampoline
-          action(rh).recoverWith {
-            case error => errorHandler.onServerError(taggedRequestHeader, error)
-          }
-        }
-        executeAction(request, taggedRequestHeader, requestBodySource, actionWithErrorHandling, errorHandler)
-
+        runAction(request, taggedRequestHeader, requestBodySource, action, errorHandler)
       case (websocket: WebSocket, Some(upgrade)) =>
-        import play.core.Execution.Implicits.trampoline
-
         val bufferLimit = config.configuration.getDeprecated[ConfigMemorySize]("play.server.websocket.frame.maxLength", "play.websocket.buffer.limit").toBytes.toInt
 
-        websocket(taggedRequestHeader).flatMap {
+        websocket(taggedRequestHeader).fast.flatMap {
           case Left(result) =>
             modelConversion.convertResult(taggedRequestHeader, result, request.protocol, errorHandler)
           case Right(flow) =>
@@ -274,15 +271,23 @@ class AkkaHttpServer(
     }
   }
 
+  @deprecated("This method is an internal API and should not be public", "2.6.10")
   def executeAction(
     request: HttpRequest,
     taggedRequestHeader: RequestHeader,
     requestBodySource: Either[ByteString, Source[ByteString, _]],
     action: EssentialAction,
-    errorHandler: HttpErrorHandler): Future[HttpResponse] = {
+    errorHandler: HttpErrorHandler): Future[HttpResponse] =
+    runAction(request, taggedRequestHeader, requestBodySource, action, errorHandler)(actorSystem.dispatcher)
 
-    import play.core.Execution.Implicits.trampoline
-    val actionAccumulator: Accumulator[ByteString, Result] = action(taggedRequestHeader)
+  private[play] def runAction(
+    request: HttpRequest,
+    taggedRequestHeader: RequestHeader,
+    requestBodySource: Either[ByteString, Source[ByteString, _]],
+    action: EssentialAction,
+    errorHandler: HttpErrorHandler)(implicit ec: ExecutionContext): Future[HttpResponse] = {
+
+    val futureAcc: Future[Accumulator[ByteString, Result]] = Future(action(taggedRequestHeader))
 
     val source = if (request.header[Expect].contains(Expect.`100-continue`)) {
       // If we expect 100 continue, then we must not feed the source into the accumulator until the accumulator
@@ -294,12 +299,18 @@ class AkkaHttpServer(
       requestBodySource
     }
 
-    val resultFuture: Future[Result] = source match {
-      case Left(bytes) if bytes.isEmpty => actionAccumulator.run()
-      case Left(bytes) => actionAccumulator.run(bytes)
-      case Right(s) => actionAccumulator.run(s)
+    // here we use FastFuture so the flatMap shouldn't actually need the executionContext
+    val resultFuture: Future[Result] = futureAcc.fast.flatMap { actionAccumulator =>
+      source match {
+        case Left(bytes) if bytes.isEmpty => actionAccumulator.run()
+        case Left(bytes) => actionAccumulator.run(bytes)
+        case Right(s) => actionAccumulator.run(s)
+      }
+    }.recoverWith {
+      case e: Throwable =>
+        errorHandler.onServerError(taggedRequestHeader, e)
     }
-    val responseFuture: Future[HttpResponse] = resultFuture.fast.flatMap { result =>
+    val responseFuture: Future[HttpResponse] = resultFuture.flatMap { result =>
       val cleanedResult: Result = resultUtils.prepareCookies(taggedRequestHeader, result)
       modelConversion.convertResult(taggedRequestHeader, cleanedResult, request.protocol, errorHandler)
     }

--- a/framework/src/play-integration-test/src/test/scala/play/it/http/HttpFiltersSpec.scala
+++ b/framework/src/play-integration-test/src/test/scala/play/it/http/HttpFiltersSpec.scala
@@ -1,0 +1,87 @@
+/*
+ * Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
+ */
+package play.it.http
+
+import play.api.http.HttpErrorHandler
+import play.api.mvc._
+import play.api.routing.Router
+import play.api.test.PlaySpecification
+import play.api.{ Application, ApplicationLoader, BuiltInComponentsFromContext, Environment }
+import play.it.test.{ ApplicationFactories, ApplicationFactory, EndpointIntegrationSpecification, OkHttpEndpointSupport }
+
+import scala.concurrent.Future
+
+class HttpFiltersSpec extends PlaySpecification
+  with EndpointIntegrationSpecification with ApplicationFactories with OkHttpEndpointSupport {
+
+  "Play http filters" should {
+
+    val appFactory: ApplicationFactory = new ApplicationFactory {
+      override def create(): Application = {
+        val components = new BuiltInComponentsFromContext(
+          ApplicationLoader.Context.create(Environment.simple())) {
+          import play.api.mvc.Results._
+          import play.api.routing.sird
+          import play.api.routing.sird._
+          override lazy val router: Router = Router.from {
+            case sird.GET(p"/") => Action { Ok("Done!") }
+            case sird.GET(p"/error") => Action { Ok("Done!") }
+            case sird.GET(p"/invalid") => Action { Ok("Done!") }
+          }
+          override lazy val httpFilters: Seq[EssentialFilter] = Seq(
+            // A non-essential filter that throws an exception
+            new Filter {
+              override def mat = materializer
+              override def apply(f: RequestHeader => Future[Result])(rh: RequestHeader): Future[Result] = {
+                if (rh.path.contains("invalid")) {
+                  throw new RuntimeException("INVALID")
+                }
+                f(rh)
+              }
+            },
+            new EssentialFilter {
+              // an essential filter returning an action that throws before returning an accumulator
+              def apply(next: EssentialAction) = EssentialAction { rh =>
+                if (rh.path.contains("error")) {
+                  throw new RuntimeException("ERROR")
+                }
+                next(rh)
+              }
+            }
+          )
+
+          override lazy val httpErrorHandler: HttpErrorHandler = new HttpErrorHandler {
+            override def onServerError(request: RequestHeader, exception: Throwable) = {
+              Future(InternalServerError(exception.getMessage))
+            }
+            override def onClientError(request: RequestHeader, statusCode: Int, message: String) = {
+              Future(InternalServerError(message))
+            }
+          }
+        }
+        components.application
+      }
+    }
+
+    "send exceptions from Filters to the HttpErrorHandler" in appFactory.withAllOkHttpEndpoints { endpoint =>
+      val request = new okhttp3.Request.Builder()
+        .url(endpoint.endpoint.pathUrl("/error"))
+        .get()
+        .build()
+      val response = endpoint.client.newCall(request).execute()
+      response.code must_== 500
+      response.body.string must_== "ERROR"
+    }
+
+    "send exceptions from EssentialFilters to the HttpErrorHandler" in appFactory.withAllOkHttpEndpoints { endpoint =>
+      val request = new okhttp3.Request.Builder()
+        .url(endpoint.endpoint.pathUrl("/invalid"))
+        .get()
+        .build()
+      val response = endpoint.client.newCall(request).execute()
+      response.code must_== 500
+      response.body.string must_== "INVALID"
+    }
+  }
+}

--- a/framework/src/play-netty-server/src/main/scala/play/core/server/netty/PlayRequestHandler.scala
+++ b/framework/src/play-netty-server/src/main/scala/play/core/server/netty/PlayRequestHandler.scala
@@ -109,13 +109,7 @@ private[play] class PlayRequestHandler(val server: NettyServer, val serverHeader
 
       //execute normal action
       case Right((action: EssentialAction, app)) =>
-        val recovered = EssentialAction { rh =>
-          import play.core.Execution.Implicits.trampoline
-          action(rh).recoverWith {
-            case error => app.errorHandler.onServerError(rh, error)
-          }
-        }
-        handleAction(recovered, requestHeader, request, Some(app))
+        handleAction(action, requestHeader, request, Some(app))
 
       case Right((ws: WebSocket, app)) if requestHeader.headers.get(HeaderNames.UPGRADE).exists(_.equalsIgnoreCase("websocket")) =>
         logger.trace("Serving this request with: " + ws)
@@ -270,19 +264,20 @@ private[play] class PlayRequestHandler(val server: NettyServer, val serverHeader
     implicit val mat: Materializer = app.fold(server.materializer)(_.materializer)
     import play.core.Execution.Implicits.trampoline
 
+    // Execute the action on the Play default execution context
+    val actionFuture = Future(action(requestHeader))(mat.executionContext)
     for {
-      bodyParser <- Future(action(requestHeader))(mat.executionContext)
-      // Execute the action and get a result
-      actionResult <- {
+      // Execute the action and get a result, calling errorHandler if errors happen in this process
+      actionResult <- actionFuture.flatMap { acc =>
         val body = modelConversion.convertRequestBody(request)
-        (body match {
-          case None => bodyParser.run()
-          case Some(source) => bodyParser.run(source)
-        }).recoverWith {
-          case error =>
-            logger.error("Cannot invoke the action", error)
-            errorHandler(app).onServerError(requestHeader, error)
+        body match {
+          case None => acc.run()
+          case Some(source) => acc.run(source)
         }
+      }.recoverWith {
+        case error =>
+          logger.error("Cannot invoke the action", error)
+          errorHandler(app).onServerError(requestHeader, error)
       }
       // Clean and validate the action's result
       validatedResult <- {

--- a/framework/src/sbt-plugin/src/main/scala-sbt-1.0/play/sbt/run/PlaySource.scala
+++ b/framework/src/sbt-plugin/src/main/scala-sbt-1.0/play/sbt/run/PlaySource.scala
@@ -1,7 +1,6 @@
 /*
  * Copyright (C) 2009-2017 Lightbend Inc. <https://www.lightbend.com>
  */
-
 // This is a naive way to make sbt.internal.io.Source accessible. That is why we
 // are declaring the package here as sbt.internal.io. You can see the code for
 // sbt.internal.io.Source here:

